conans/test/remote_checks_test.py

```diff
@@ -0,0 +1,138 @@
+        conanfile = """from conans import ConanFile
+class Pkg(ConanFile):
+    pass"""
+        client.save({"conanfile.py": conanfile})
+        client.run("create . Pkg/0.1@lasote/testing")
+        client.run("upload Pkg* --all -r=server1 --confirm")
+        client.run("upload Pkg* --all -r=server2 --confirm")
+
+        # It takes the default remote
+        client.run("remove * -f")
+        client.run("remote list_ref")
+        self.assertNotIn("Pkg", client.out)
+
+        client.run("export . Pkg/0.1@lasote/testing")
+        client.run("install Pkg/0.1@lasote/testing")
+        self.assertIn("Downloading conan_package.tgz", client.out)
+        client.run("remote list_ref")
+        self.assertIn("Pkg/0.1@lasote/testing: server1", client.out)
+
+        # Explicit remote also defines the remote
+        client.run("remove * -f")
+        client.run("remote list_ref")
+        self.assertNotIn("Pkg", client.out)
+        client.run("export . Pkg/0.1@lasote/testing")
+        client.run("install Pkg/0.1@lasote/testing -r=server2")
+        self.assertIn("Downloading conan_package.tgz", client.out)
+        client.run("remote list_ref")
+        self.assertIn("Pkg/0.1@lasote/testing: server2", client.out)
+
+        # But order fails!!!
+        client.run("remove * -f")
+        client.run("remove * -f -r=server1")
+        client.run("export . Pkg/0.1@lasote/testing")
+        error = client.run("install Pkg/0.1@lasote/testing", ignore_error=True)
+        self.assertTrue(error)
+        self.assertIn("Missing prebuilt package for 'Pkg/0.1@lasote/testing'", client.out)
+        client.run("remote list_ref")
+        self.assertNotIn("Pkg", client.out)
+
+    def test_binaries_from_different_remotes(self):
+        servers = {"server1": TestServer(), "server2": TestServer(), "server3": TestServer()}
+        client = TestClient(servers=servers, users={"server1": [("lasote", "mypass")],
+                                                    "server2": [("lasote", "mypass")]})
+        conanfile = """from conans import ConanFile
+class Pkg(ConanFile):
+    options = {"opt": [1, 2, 3]}
+"""
+        client.save({"conanfile.py": conanfile})
+        client.run("create . Pkg/0.1@lasote/testing -o Pkg:opt=1")
+        client.run("upload Pkg* --all -r=server1 --confirm")
+        client.run("remove * -p -f")
+        client.run("create . Pkg/0.1@lasote/testing -o Pkg:opt=2")
+        client.run("upload Pkg* --all -r=server2 --confirm")
+        client.run("remove * -p -f")
+        client.run("remote list_ref")
+        self.assertIn("Pkg/0.1@lasote/testing: server1", client.out)
+        # Trying to install from another remote fails
+        error = client.run("install Pkg/0.1@lasote/testing -o Pkg:opt=2 -r=server2", ignore_error=True)
+        self.assertTrue(error)
+        self.assertIn("ERROR: Missing prebuilt package for 'Pkg/0.1@lasote/testing'", client.out)
```

Comment:
If it is the desired behavior to only use remote1, we should raise because of the "-r server2"

Reply:
The problem is that it would be breaking for dependencies not installed which will be retrieved from server2.

Reply:
I know probably it is a nightmare, but better than current behavior would be to: 

1. Remove the recipe because it is from remote1 and we don't have a package.
2. Install from remote2, the Pkg and all the dependencies (that have missing recipes/packages in the local cache)

If any of them is not present, it should fail.

Reply:
In other words: If I specify a remote I want:

1. To use the recipes and packages that are present in the local cache.
2. If a recipe or a package is not in the local cache, gather it from the specified remote.